core/src/main/scala/org/bitcoins/core/wallet/builder/TxBuilder.scala

```diff
@@ -244,167 +256,147 @@ sealed abstract class BitcoinTxBuilder extends TxBuilder {
-              Future.fromTry(TxBuilderError.NestedWitnessSPK)
-            case _: P2SHScriptPubKey =>
-              Future.fromTry(TxBuilderError.NestedP2SHSPK)
-            case _: NonStandardScriptPubKey | _: WitnessCommitment |
-                EmptyScriptPubKey | _: UnassignedWitnessScriptPubKey =>
-              Future.fromTry(TxBuilderError.NoSigner)
+        case p2shInfo: P2SHNoNestSpendingInfo =>
+          signAndAddInputForP2SH(p2shInfo,
+                                 unsignedTx,
+                                 oldInput.sequence,
+                                 inputIndex.toInt,
+                                 dummySignatures)
+        case p2shNestedSegwitInfo @ P2SHNestedSegwitV0UTXOSpendingInfo(
+              _,
+              amount,
+              scriptPubKey,
+              _,
+              hashType,
+              redeemWitnessScript,
+              _) =>
+          redeemWitnessScript match {
+            case _: P2WSHWitnessSPKV0 =>
+              signAndAddInputForP2SH(p2shNestedSegwitInfo,
+                                     unsignedTx,
+                                     oldInput.sequence,
+                                     inputIndex.toInt,
+                                     dummySignatures)
+            case p2wpkh: P2WPKHWitnessSPKV0 =>
+              val uwtx = WitnessTransaction.toWitnessTx(unsignedTx)
+
+              //breaks an abstraction inside of all of the signers
+              //won't be able to be handled properly until gemini stuff
+              //is open sourced
+              signP2SHP2WPKH(
+                unsignedTx = uwtx,
+                inputIndex = inputIndex,
+                output = TransactionOutput(amount, scriptPubKey),
+                p2wpkh = p2wpkh,
+                utxo = utxo,
+                hashType = hashType,
+                dummySignatures = dummySignatures
+              )
           }
-          sigComponentF.map(_.transaction)
-        case _: NonStandardScriptPubKey | _: WitnessCommitment |
-            EmptyScriptPubKey | _: UnassignedWitnessScriptPubKey =>
+        case _: UnassignedSegwitNativeUTXOSpendingInfo =>
           Future.fromTry(TxBuilderError.NoSigner)
       }
     }
   }
 
+  private def signAndAddInputForP2SH(
+      spendingInfo: P2SHSpendingInfo,
+      unsignedTx: Transaction,
+      sequence: UInt32,
+      inputIndex: Int,
+      dummySignatures: Boolean)(
+      implicit ec: ExecutionContext): Future[Transaction] = {
+    require(!spendingInfo.redeemScript.isInstanceOf[P2WPKHWitnessSPKV0],
+            "Should call signP2SHP2WPKH")
+
+    val outpoint = spendingInfo.outPoint
+    val redeemScript = spendingInfo.redeemScript
+
+    val input =
+      TransactionInput(outpoint, EmptyScriptSignature, sequence)
+
+    val updatedTx =
+      unsignedTx.updateInput(inputIndex, input)
+
+    val updatedOutput =
+      TransactionOutput(spendingInfo.amount, redeemScript)
+
+    val updatedUTXOInfo = BitcoinUTXOSpendingInfo(outpoint,
```

Comment:
Do we need to call this here? I would like to get rid of this if possible and directly call constructors on the subtype of `BitcoinUTXOSpendingInfo()`. I guess that would force a pattern match here then though?

Reply:
I think that this is the correct use of `BitcoinUTXOSpendingInfo.apply`, what would go here if I didn't use this would be exactly the match and construction defined in that method for reuse. I think there are times when you have a `ScriptPubKey` on your hands that you have the spending info for but don't have the elaborated type, and that is what the `apply` is for.